src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/helpers/AuthorizationRequestedHelper.kt

```diff
@@ -0,0 +1,135 @@
+import it.pagopa.ecommerce.eventdispatcher.services.eventretry.v2.AuthorizationStateRetrieverRetryService
+import it.pagopa.ecommerce.eventdispatcher.services.v2.AuthorizationStateRetrieverService
+import it.pagopa.ecommerce.eventdispatcher.utils.DeadLetterTracedQueueAsyncClient
+import java.util.*
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+
+data class AuthorizationRequestedEvent(
+  val requested: QueueEvent<TransactionAuthorizationRequestedEvent>?,
+  val retried: QueueEvent<TransactionAuthorizationOutcomeWaitingEvent>?
+) {
+  init {
+    require(listOfNotNull(requested, retried).size == 1) { "Only one event must be non-null!" }
+  }
+
+  companion object {
+    fun requested(
+      event: QueueEvent<TransactionAuthorizationRequestedEvent>
+    ): AuthorizationRequestedEvent = AuthorizationRequestedEvent(event, null)
+
+    fun retried(
+      event: QueueEvent<TransactionAuthorizationOutcomeWaitingEvent>
+    ): AuthorizationRequestedEvent = AuthorizationRequestedEvent(null, event)
+  }
+
+  fun <T> fold(
+    onRequested: (QueueEvent<TransactionAuthorizationRequestedEvent>) -> T,
+    onRetried: (QueueEvent<TransactionAuthorizationOutcomeWaitingEvent>) -> T,
+  ): T {
+    return checkNotNull(
+      when {
+        requested != null -> onRequested(requested)
+        retried != null -> onRetried(retried)
+        else -> null
+      }) { "No variant of `AuthorizationRequestedEvent` is non-null!" }
+  }
+}
+
+/**
+ * This helper implements the business logic related to handling calling `getState` from NPG. In
+ * particular, the [getAuthorizationState] method does the following:
+ * - checks for the transaction current status
+ * - determines whether the transaction was requesting authorization via NPG
+ * - calls NPG's `getSTate`
+ * - calls transactions-service PATCH auth-request
+ * - enqueues a retry event in case of error
+ */
+@Component
+class AuthorizationRequestedHelper(
+  @Autowired private val transactionsServiceClient: TransactionsServiceClient,
+  @Autowired private val transactionsEventStoreRepository: TransactionsEventStoreRepository<Any>,
+  @Autowired
+  private val authorizationStateRetrieverRetryService: AuthorizationStateRetrieverRetryService,
+  @Autowired private val authorizationStateRetrieverService: AuthorizationStateRetrieverService,
+  @Autowired private val deadLetterTracedQueueAsyncClient: DeadLetterTracedQueueAsyncClient,
+  @Autowired private val tracingUtils: TracingUtils,
+  @Autowired private val strictSerializerProviderV2: StrictJsonSerializerProvider
+) {
+
+  var logger: Logger = LoggerFactory.getLogger(AuthorizationRequestedHelper::class.java)
+
+  fun authorizationStateRetrieve(
+    parsedEvent: AuthorizationRequestedEvent,
```

Comment:
see above comment, here can be used Either data structure: is well known and more readable than a custom class (and in this situation there are only two variant)